crates/cli-support/src/wit/mod.rs

```diff
@@ -484,7 +484,30 @@ impl<'a> Context<'a> {
             Some(class) => {
                 let class = class.to_string();
                 match export.method_kind {
-                    decode::MethodKind::Constructor => AuxExportKind::Constructor(class),
+                    decode::MethodKind::Constructor => {
+                        let msg = "Trying to return JS primitive from constructor, return value will be ignored. Use a builder instead (remove `constructor` attribute).";
+                        match descriptor.ret {
+                            Descriptor::I8
+                            | Descriptor::U8
+                            | Descriptor::ClampedU8
+                            | Descriptor::I16
+                            | Descriptor::U16
+                            | Descriptor::I32
+                            | Descriptor::U32
+                            | Descriptor::F32
+                            | Descriptor::F64
+                            | Descriptor::I64
+                            | Descriptor::U64
+                            | Descriptor::Boolean
+                            | Descriptor::Char
+                            | Descriptor::CachedString
+                            | Descriptor::String
+                            | Descriptor::Option(_)
+                            | Descriptor::Unit => bail!(msg),
```

Comment:
This should also include `Enum`, since (exported) enums are represented as integers in JS.

`Ref`s and `RefMut`s to any of these should probably also be included, since they'll still map to primitives while behind a reference, although it doesn't matter a whole lot since you can't return references right now anyway.

(Imported / string enums are handled purely at the macro level, so we don't need to worry about them here.)

Reply:
I moved this check to it's own function so that it is cleaner to check nested descriptors. Like you mentioned `Ref` and `RefMut` are not relevant for now, but it occurred to me that `Result` still is.